### ISSUE TITLE
`MaxDatagramFrameSize` should be configurable

### ISSUE THREAD

**Issue by geokb** [quic-go/quic-go#3599]
Hi @lucas-clemente and thank you for your tremendous effort on bringing QUIC to Go. What I am trying to do is basically something similar to the functionality of WireGuard but over QUIC datagrams (i.e. sending IP packets over datagrams between 2 TUN devices). The problem is that `MaxDatagramFrameSize` is hardwired to `1220` and even recently `1200` in a merged PR however you must have at least an MTU of `1280` to be able to fit a full IPv6 packet inside a single QUIC datagram since 1280 is the smallest possible MTU for the TUN device (or any type of link actually) in the case of IPv6. In other words, you cannot create a QUIC-based tunnel that can carry IPv6 packets unless the payload is coincidentally small enough to fit inside `MaxDatagramFrameSize` otherwise the packet is dropped and the result is that the application communicating over the tunnel is not communicating at all (i.e. `conn.SendMessage(buf)` returns an error with msg `message too large`).
In fact your hardwired value of `MaxDatagramFrameSize` also equally affects IPv4 traffic unless the MTU of the TUN device is set to a lower value than `MaxDatagramFrameSize` but thankfully in this case IPv4 can afford to use MTUs as low as `576` even though this obviously hurts performance.

To be quite honest I have no idea why this value is hardwired and hidden from the developer in the first place. I'd like also to mention that my case uses a single datagram fully for a single IPv4 or IPv6 packet. However this might not the case in other applications, let's say you put the IP packet inside a bigger struct that has more info than just the IP packet (e.g. ids used for demuxing over multiple links) and is serialized into a protobuf stream, and that protobuf serialized data is what has to fit inside the `MaxDatagramFrameSize`. So in such case, even hardwiring this value to 1280 which is enough to fit any IPv6 packet is still  not enough to fit for the protobuf data.

TLDR; I totally believe that this constant should be exposed publicly inside `quic.Config` and be left up to the developer and the application to decide what its value should be.

**Comment by marten-seemann** [comment]
The problem is that just configuring a large value doesn’t work. You also need to be able to send QUIC packets that large on the outer connection, and that’s only possible after a few rounds of DPLPMTUD, if at all.

**Comment by geokb** [comment]
@marten-seemann thanks for the response. So you're basically hardcoding a small enough value for `MaxDatagramFrameSize` in order for the overall QUIC packet to fit inside at least an outer MTU of 1280 (i.e. 1232 QUIC packet size for IPv6)? So you're basically assuming that the outer MTU is 1280. I get it but should not the developer be the one who gets to play with that value depending on the application? Most outer MTUs are far greater than 1280 and usually are bigger than 1400 if not close to 1500. Your hardcoded value basically kills any chance of developing many if not most real world tunneling applications since it only works for IPv4 tunneling where the inner MTU must be equal or less than `MaxDatagramFrameSize`. I guess you could still use your value as a default or fallback while giving the developer the freedom to change it if they have to. Let me state some of the consequences of your hardcoded value again:

1. IPv6 tunneling is simply impossible.
2. IPv4 tunneling is also impossible unless the application sets the inner MTU to `MaxDatagramFrameSize` or less and makes sure to change it whenever you decide to change it. If you decrease it for some other reason in the future, I must know of it and I must either decrease immediately it or not be able to update the quic-go package at all.
3. 1 & 2 even assume that the datagram is only transporting raw IP packets. However, I could want to develop an application where I can add my own application-layer overhead that must fit inside the datagram (e.g. protobuf, flatbuffers, etc...). afaik MASQUE and WebTransport which use HTTP3 datagrams add their own overhead that must fit inside the underlying QUIC datagram. So if you're trying to tunnel IPv6 packets inside MASQUE it's also impossible unless you calculate for the overhead used by the h3 datagram.
4. QUIC inside IP packet tunneled over your implementation of QUIC is also impossible (e.g. the inner QUIC won't fit unless the inner MTU which used by the tunnel is at least 1280)

I get that you want for the quic-go to universally work so you're hardcoding for the strictest case possible. However, 1280 is rarely used in real world cases unless by some VPNs. I'd appreciate it if you expose this value in `quic.Config` while fallbacking to your value if not explicitly set. This will maintain the current behavior while giving the chance to application devs to build serious and complex tunneling applications.

**Comment by marten-seemann** [comment]
I’m aware of all of this, you don’t need to convince me that it would be nice to be able to send larger datagram frames.

This is an API problem. Feel free to suggest an API that makes that work. I don’t have the bandwidth to work on this any time soon.

**Comment by marten-seemann** [comment]
One option would be to put this on the `ConnectionState`. When DPLPMTUD discovers a higher MTU, this value would be updated. The application could regularly call `ConnectionState` if it wishes to send larger datagrams.

**Comment by marten-seemann** [comment]
Fixed by #4143.
